ra/ra.go

```diff
@@ -2197,55 +2197,73 @@ func (ra *RegistrationAuthorityImpl) AdministrativelyRevokeCertificate(ctx conte
 	}
 
 	reasonCode := revocation.Reason(req.Code)
-	if reasonCode == ocsp.KeyCompromise && req.Cert == nil && !req.SkipBlockKey {
-		return nil, fmt.Errorf("cannot revoke and block for KeyCompromise by serial alone")
```

Comment:
It seems like this error case is lost in the refactoring, but I think it's still a useful one.

Reply:
It was purposefully removed because the semantics we actually want are (imo) "can't revoke for key compromise if we don't have access to the corresponding pubkey to block". So that's now covered by "cannot revoke malformed certificate for KeyCompromise".

Reply:
I thought part of the idea here was that if there's a key compromise we should have more evidence at hand than the serial number; either a private key or a public key (with corresponding self-signed something). Wouldn't it be best to require one of those as input to avoid operator errors? In particular I'm thinking of something like this: https://blog.hboeck.de/archives/888-How-I-tricked-Symantec-with-a-Fake-Private-Key.html

Of course that check doesn't have to happen at the RA; it could also happen in admin-revoker.

Reply:
We don't enforce that today -- an authorized person can run `admin-revoker malformed-revoke deadbeef  1` and it will trust the operator's judgement --  and my goal with these three PRs is to preserve current functionality as much as possible.

The `admin` PR stacked on top of this does have modes of operation where it takes a private key as input. If we want, in that PR or a follow-up we can restrict the `keyCompromise` revocation reason to that operating mode. (Although I have concerns about that, as it wouldn't have worked in the most recent case where we were provided with a pubkey and a self-signed thingy, and automating verifying every different kind of self-signed thingy someone might provide is exactly why we have the ACME revocation API endpoint.)

In the example you link, Symantec _was_ provided with a (malformed) private key, so requiring that the tool only operate on top of private keys and not serials does not in and of itself prevent that.

Reply:
I think what we want here is for the mainstream methods of keyCompromise revocation to expect a private or public key, and have an emergency trapdoor that revokes by serial for keyCompromise, which is only intended for highly exceptional use.

> In the example you link, Symantec was provided with a (malformed) private key, so requiring that the tool only operate on top of private keys and not serials does not in and of itself prevent that.

They were provided with a private key, but programmatic checks would have noted it didn't match the purported public key and would have blocked revocation. My goal is that the mainstream paths for keyCompromise revocation should do that sort of verification automatically.

Reply:
I forgot to put it in the comment above, but I'm convinced that this case (in the RA) can go away, though we should be thinking about the issues of requiring public or private key in terms of the user-facing interface of admin-revoker.